Documentation/tutorial/intro markdown lite.md

```diff
@@ -0,0 +1,199 @@
+* [Render](xref:Microsoft.DocAsCode.MarkdownLite.IMarkdownRenderer)
+
+### Step 1: Parse
+
+In this step, it will parse markdown text to [tokens](xref:Microsoft.DocAsCode.MarkdownLite.IMarkdownToken).
+The parser is based on [rules](xref:Microsoft.DocAsCode.MarkdownLite.IMarkdownRule), which make up the [context](xref:Microsoft.DocAsCode.MarkdownLite.IMarkdownContext).
+
+For example,
+[heading token](xref:Microsoft.DocAsCode.MarkdownLite.MarkdownHeadingBlockToken) is created by [heading rule](xref:Microsoft.DocAsCode.MarkdownLite.MarkdownHeadingBlockRule),
+the heading rule is belonging to [block context](xref:Microsoft.DocAsCode.MarkdownLite.MarkdownBlockContext).
+
+### Step 2: Rewrite or validate
+
+In this step, it will walk through all [models](xref:Microsoft.DocAsCode.MarkdownLite.IMarkdownToken),
+we can change it to another, or just validate the content.
+
+For example, we can create a rewriter to change all heading token with depth + 1:
+
+```csharp
+MarkdownTokenRewriterFactory.FromLambda<IMarkdownRewriteEngine, MarkdownHeadingBlockToken>(
+    (engine, token) => new MarkdownHeadingBlockToken(token.Rule, token.Context, token.Content, token.Id, token.Depth + 1, token.SourceInfo);
+```
+
+### Step 3: Render
+
+In this step, it render models to text content (html format for normal).
+To simplify extension, we create an [adapter](xref:Microsoft.DocAsCode.MarkdownLite.MarkdownRendererAdapter),
+the adapter invoke methods by following rules:
+
+1.  Method name is `Render`
+2.  Instance method
+3.  Return type is @Microsoft.DocAsCode.MarkdownLite.StringBuffer
+4.  The count of parameters is 3, and types are following:
+    1.  @Microsoft.DocAsCode.MarkdownLite.IMarkdownRenderer or any type implements it.
+    2.  @Microsoft.DocAsCode.MarkdownLite.IMarkdownToken or any type implements it.
+    3.  @Microsoft.DocAsCode.MarkdownLite.IMarkdownContext or any type implements it.
+5.  Alway invoke the most match method.
+
+## Engine and engine builder
+
+Engine is a set of parser, rewriter and renderer.
+It can markup a markdown file to html file (or others).
+But it cannot be invoked in parallel.
+
+So we create an [engine builder](xref:Microsoft.DocAsCode.MarkdownLite.MarkdownEngineBuilder).
+It defines all the rules of parser, rewriter and renderer.
+It can create instances when needed.
+
+## How to customize markdown syntax
+
+### Define markdown syntax
+
+Define markdown:
+
+```
+: My label
+```
+
+should be rendered as following html:
+
+```html
+<div id="My label"></div>
+```
+
+### Select context
+
+First of all, we should select the context for this rule.
+And in this goal, the new line is required.
+So it should be block context, all of the names for class should contain `Block`.
+
+> [!TIP]
+> In markdown, there is two context:
+> * block
+> * inline
+>
+> Block rules require a new line in the end.
```

Comment:
I think markdown spec itself has a definition of block and inline tokens:
[https://daringfireball.net/projects/markdown/syntax#block](https://daringfireball.net/projects/markdown/syntax#block)